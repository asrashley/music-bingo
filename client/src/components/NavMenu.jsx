import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { reverse } from 'named-urls';
import { Link } from 'react-router-dom';

import { routes } from '../routes/routes';

import { MenuItemPropType, SectionItemPropType } from '../types/Menu';
import { UserPropType } from '../user/types/User';

export function DropDownMenuItem({ href, title, onClick }) {
    return (
        <li>
            <Link className="dropdown-item" onClick={onClick} to={href}>{title}</Link>
        </li>
    );
}
DropDownMenuItem.propTypes = {
    href: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    onClick: PropTypes.func.isRequired,
};

export function DropDownMenu({ title, items, section }) {
    const [expanded, setExpanded] = useState(false);
    const show = expanded ? 'show' : '';

    const toggleExpand = () => setExpanded(!expanded);

    return (
        <li className={`nav-item dropdown ${section.item ?? ''}`}>
            <button className={`nav-link dropdown-toggle ${section.link} ${show}`}
                aria-expanded={expanded} onClick={toggleExpand}>
                {title}
            </button>
            <ul className={`dropdown-menu  ${show}`}>
                {items.map(({ title, href }, idx) => <DropDownMenuItem
                    key={`${idx}-${title}`} title={title} href={href} onClick={() => setExpanded(false)} />)}
            </ul>
        </li>
    );
}
DropDownMenu.propTypes = {
    section: SectionItemPropType.isRequired,
    title: PropTypes.string.isRequired,
    items: PropTypes.arrayOf(MenuItemPropType),
};

export function NavMenuComponent({
    className, gameMenu, sections, user, userMenu }) {
    return (
        <nav id="nav-menu" className={className}>
            <Link className="navbar-brand" to={reverse(`${routes.index}`)}>
                <img src={`${process.env.PUBLIC_URL}/img/Icon.png`} width="30" height="30"
                    className="d-inline-block align-top" alt="" />
                Musical Bingo
            </Link>
            <button className="navbar-toggler" type="button"
                data-toggle="collapse" data-target="#nav-menu"
                aria-controls="nav-menu" aria-expanded="false"
                aria-label="Toggle navigation">
                <span className="navbar-toggler-icon"></span>
            </button>
            <div className="collapse navbar-collapse" id="navbarNav">
                <ul className="nav nav-pills">
                    <li className={`nav-item ${sections.Home.item}`}>
                        <Link className={`nav-link ${sections.Home.link}`}
                            to={reverse(`${routes.index}`)}>
                            Home <span className="sr-only">(current)</span>
                        </Link>
                    </li>
                    <li className={`nav-item ${sections.Game.item}`}>
                        <Link className={`nav-link ${sections.Game.link}`}
                            to={reverse(`${routes.listGames}`)}
                        >Now Playing
                        </Link>
                    </li>
                    <DropDownMenu
                        title="Previous Games"
                        items={gameMenu}
                        section={sections.History}
                    />
                    {user.groups.creators && <li className={`nav-item ${sections.Clips.item}`}>
                        <Link className={`nav-link ${sections.Clips.link}`}
                            to={reverse(`${routes.listDirectories}`)}
                        >Clips
                        </Link>
                    </li>}
                    {user.loggedIn ? <DropDownMenu
                        title={user.loggedIn ? user.username : 'Log In'}
                        items={userMenu}
                        section={sections.User}
                    /> : <li className={`nav-item ${sections.User.item}`}>
                        <Link to={reverse(`${routes.login}`)}
                            className={`login nav-link  ${sections.User.link}`}
                        >Log in</Link>
                    </li>
                    }
                </ul>
            </div>
        </nav >
    );
}

NavMenuComponent.propTypes = {
    className: PropTypes.string.isRequired,
    gameMenu: PropTypes.arrayOf(MenuItemPropType),
    sections: PropTypes.objectOf(SectionItemPropType),
    user: UserPropType.isRequired,
    userMenu: PropTypes.arrayOf(MenuItemPropType),
};
