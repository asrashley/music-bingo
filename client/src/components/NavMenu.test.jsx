import { fireEvent } from "@testing-library/dom";
import { screen } from '@testing-library/react'
import { renderWithProviders } from "../../tests";

import { DropDownMenuItem, DropDownMenu, NavMenuComponent } from './NavMenu';

describe('NavMenuComponent', () => {
    const gameMenu = [{
        title: 'game title one',
        href: '/game/one',

    }, {
        title: 'game title two',
        href: '/game/two',
    }];
    const userMenu = [{
        title: 'User title one',
        href: '/user/one',

    }, {
        title: 'user title two',
        href: '/user/two',
    }];
    const notActiveItem = (id) => ({
        id,
        active: false,
        item: '',
        link: '',
    });
    const sections = {
        Home: notActiveItem('Home'),
        Game: notActiveItem('Game'),
        Clips: notActiveItem('Clips'),
        History: {
            id: 'History',
            active: true,
            item: 'active',
            link: 'active',
        },
        User: notActiveItem('User'),
    };

    it('DropdownMenuItem', async () => {
        const title = 'Menu Item Title';
        const { container, events, getByText } = renderWithProviders(
            <DropDownMenuItem href="/href-link" title={title} />);
        expect(container.querySelector('a[href="/href-link"]')).not.toBeNull();
        await events.click(getByText(title));
    });

    it('DropDownMenu', async () => {
        const title = "Menu Title";
        const section = {
            id: 'test',
            active: true,
            item: 'active',
            link: 'active',
        };
        const { events, getByText, findAllByText } = renderWithProviders(
            <DropDownMenu items={gameMenu} section={section} title={title} />);
        getByText(title);
        gameMenu.forEach(item => {
            getByText(item.title);
        })
        fireEvent.click(getByText(title));
        await findAllByText(title);
        await events.click(screen.getByText(gameMenu[1].title));
        await findAllByText(title);
        expect(document.querySelector('.dropdown-menu.show')).toBeNull();
    });

    describe('NavMenuComponent', () => {
        it('matches snapshot', () => {
            const user = {
                pk: 5,
                loggedIn: true,
                email: 'Fred.Flintstone@unit.test',
                username: 'Fred Flintstone',
                groups: {},
            };
            const props = {
                className: 'menu-class',
                sections,
                user,
                gameMenu,
                userMenu,
            };
            const { asFragment, getByText, queryByText } = renderWithProviders(<NavMenuComponent {...props} />);
            getByText(user.username);
            expect(queryByText('Clips')).toBeNull();
            expect(asFragment()).toMatchSnapshot();
        });

        it('shows clip menu for users in the creators group', () => {
            const user = {
                loggedIn: true,
                username: 'a.user',
                pk: 6,
                email: 'a.user@unit.test',
                groups: {
                    users: true,
                    creators: true,
                },
            };
            const props = {
                className: 'menu-class',
                sections,
                user,
                gameMenu,
                userMenu,
            };
            const { getByText } = renderWithProviders(<NavMenuComponent {...props} />);
            getByText(user.username);
            getByText('Clips');
        });

        it('shows log in invite', () => {
            const user = {
                loggedIn: false,
                username: 'not.shown',
                pk: 8,
                email: '',
                groups: {
                },
            };
            const props = {
                className: 'menu-class',
                sections,
                user,
                gameMenu,
                userMenu,
            };
            const { getByText, queryByText } = renderWithProviders(<NavMenuComponent {...props} />);
            expect(queryByText('Clips')).toBeNull();
            expect(queryByText('not.shown')).toBeNull();
            getByText('Log in');
        });
    });
});