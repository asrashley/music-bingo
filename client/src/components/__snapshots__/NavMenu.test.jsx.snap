// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavMenuComponent > NavMenuComponent > matches snapshot 1`] = `
<DocumentFragment>
  <nav
    class="menu-class"
    id="nav-menu"
  >
    <a
      class="navbar-brand"
      href="/"
    >
      <img
        alt=""
        class="d-inline-block align-top"
        height="30"
        src="/public/img/Icon.png"
        width="30"
      />
      Musical Bingo
    </a>
    <button
      aria-controls="nav-menu"
      aria-expanded="false"
      aria-label="Toggle navigation"
      class="navbar-toggler"
      data-target="#nav-menu"
      data-toggle="collapse"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="collapse navbar-collapse"
      id="navbarNav"
    >
      <ul
        class="nav nav-pills"
      >
        <li
          class="nav-item "
        >
          <a
            class="nav-link "
            href="/"
          >
            Home 
            <span
              class="sr-only"
            >
              (current)
            </span>
          </a>
        </li>
        <li
          class="nav-item "
        >
          <a
            class="nav-link "
            href="/game"
          >
            Now Playing
          </a>
        </li>
        <li
          class="nav-item dropdown active"
        >
          <button
            aria-expanded="false"
            class="nav-link dropdown-toggle active "
          >
            Previous Games
          </button>
          <ul
            class="dropdown-menu  "
          >
            <li>
              <a
                class="dropdown-item"
                href="/game/one"
              >
                game title one
              </a>
            </li>
            <li>
              <a
                class="dropdown-item"
                href="/game/two"
              >
                game title two
              </a>
            </li>
          </ul>
        </li>
        <li
          class="nav-item dropdown "
        >
          <button
            aria-expanded="false"
            class="nav-link dropdown-toggle  "
          >
            Fred Flintstone
          </button>
          <ul
            class="dropdown-menu  "
          >
            <li>
              <a
                class="dropdown-item"
                href="/user/one"
              >
                User title one
              </a>
            </li>
            <li>
              <a
                class="dropdown-item"
                href="/user/two"
              >
                user title two
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </div>
  </nav>
</DocumentFragment>
`;
